pyteal/compiler/scratchslots.py

```diff
@@ -2,45 +2,99 @@
 
 from ..ast import ScratchSlot, SubroutineDefinition
 from ..ir import Mode, TealComponent, TealBlock
+from ..ir.ops import Op
 from ..errors import TealInputError, TealInternalError
 from ..config import NUM_SLOTS
 
 
+def collectUnoptimizedSlotIDs(
+    subroutineBlocks: Dict[Optional[SubroutineDefinition], TealBlock]
+) -> Set[int]:
+    """Find and return all referenced ScratchSlots that need to be skipped
+    during optimization.
+
+    Args:
+        subroutineBlocks: A mapping from subroutine to the subroutine's control flow graph.
+        The key None is taken to mean the main program routine.
+
+    Returns:
+        A set which contains the slot ids used by DynamicScratchVars, all the reserved
+        slot ids, and all global slots.
+    """
+
+    unoptimized_slots: Set[int] = set()
+
+    def collectSlotsFromBlock(block: TealBlock):
+        for op in block.ops:
+            for slot in op.getSlots():
+                # dynamic slot or reserved slot
+                if op.op == Op.int or slot.isReservedSlot:
+                    unoptimized_slots.add(slot.id)
```

Comment:
I think it would be better to return `Set[ScratchSlot]` instead of `Set[int]`, since the integer ID is really an implementation detail of how slot equality is determined.